### PR TITLE
[Feature] Add option to pass a raw javascript object to set and add methods

### PR DESCRIPTION
**Pull request by netochaves**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Have you added new tests to prevent regressions?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you update the typescript typings accordingly (if applicable)?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/master/CONTRIBUTING.md)?

### Description of change
#10397
Before it isn't possible to pass just a raw javascript object without IDs to set method

When calling a custom set method of `has-many` association you can pass just a javascript object with no need of a key, e.g: ` setProductImages(req.body.productsImages)`

### CODE REVIEW

**Comment by papb** [comment]
Hello, I see you managed to open the PR, after we talked a bit! That's great! Thank you for taking the time to help Sequelize! I hope to see more PRs from you in the future!

I will now review your PR.

**Comment by netochaves** [comment]
Thank you for you help, waiting for your review :)

**Comment by netochaves** [comment]
In docker all tests works fine locally, i test with mysql and mssql, i'll investigate it.

File: lib/associations/base.js [hunk 1]
```diff
@@ -109,6 +109,31 @@ class Association {
...
 
+  /**
+   * Normalize input and create the model if necessary
+   *
+   * @param {Array|string} input it may be array or single obj, instance or primary key
+   *
+   * @private
+   * @returns {Array} built or create objects
+   */
+  toInstanceCreate(input) {
```
Comment (marclerodrigues) [inline lib/associations/base.js]: Specs?

**Comment by papb** [comment]
Hello, I am having trouble to understand your code to be honest. What is the idea behind this `toInstanceCreate` method you created?

**Comment by netochaves** [comment]
The idea is that when using `set` and `add` methods we use the `toInstanceCreate` to create the instance if the parameter is just a array of objects. If not he‘ll have the same behavior of `toInstanceArray`. I’ve created this one because toInstanceArray is being used on another places.

**Comment by papb** [comment]
@netochaves Sorry to take long. I added another review, and I would also like you to add more complex tests, such as tests for `add` and `set` when there were already some other records associated, where only some of the elements in the array are raw and others are full instances, etc

File: test/integration/associations/has-many.test.js [hunk 2]
```diff
@@ -1265,7 +1314,7 @@ describe(Support.getTestDialectTeaser('HasMany'), () => {
               Task.create({ title: 'task' })
             ]);
           }).then(([user, task]) => {
-            return user.setTasks([task]).return(user);
+            return Promise.resolve(user.setTasks([task])).return(user);
```
[the remark this answers is not in the document]
Comment (netochaves) [reply test/integration/associations/has-many.test.js]: Actually it's not necessary, i'll change it.

File: test/integration/associations/has-many.test.js [hunk 3]
```diff
@@ -1321,7 +1370,7 @@ describe(Support.getTestDialectTeaser('HasMany'), () => {
...
```
[the remark this answers is not in the document]
Comment (netochaves) [reply test/integration/associations/has-many.test.js]: Actually it's not necessary, i'll change it.

File: lib/associations/base.js [hunk 4]
```diff
@@ -109,6 +109,31 @@ class Association {
...
+    if (!Array.isArray(input)) {
+      input = [input];
+    }
+    return Promise.all(input.map(element => {
```
[the remark this answers is not in the document]
Comment (netochaves) [reply lib/associations/base.js]: Nice!

File: lib/associations/has-many.js [hunk 5]
```diff
@@ -397,18 +408,19 @@ class HasMany extends Association {
...
-    Object.assign(update, this.scope);
-
-    const where = {
-      [this.target.primaryKeyAttribute]: targetInstances.map(unassociatedObject =>
-        unassociatedObject.get(this.target.primaryKeyAttribute)
-      )
-    };
-
-    return this.target.unscoped().update(update, _.defaults({ where }, options)).return(sourceInstance);
+    return this.toInstanceCreate(targetInstances).then(resulInstances => {
```
[the remark this answers is not in the document]
Comment (netochaves) [reply lib/associations/has-many.js]: i'll fix it.

File: test/integration/associations/base.test.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+      this.Product.hasMany(this.ProductImages);
+
+      return current.sync({ force: true });
+    });
+
+    it('Should receive a array of models from toInstanceCreate', () => {
+      const productImages = [
+        { image: 'image1', thunbmail: 'thunbmail1' },
+        { image: 'image1', thunbmail: 'thunbmail1' }
```
[the remark this answers is not in the document]
Comment (netochaves) [reply test/integration/associations/base.test.js]: ok thanks

File: test/integration/associations/has-many.test.js [hunk 7]
```diff
@@ -933,6 +933,55 @@ describe(Support.getTestDialectTeaser('HasMany'), () => {
...
+          return product.setProductImages([
+            { image: 'test1', thumbnail: 'test1' },
+            { image: 'test2', thumbnail: 'test2' }
+          ]);
+        }).then(() => {
+          return expect(ProductImage.findAll()).to.eventually.have.length(2);
+        }).then(() => {
+          return Product.findOne({ include: ProductImage }).then(product => {
+            expect(product.ProductImages).to.be.an('array');
+            expect(product.ProductImages).to.have.length(2);
```
[the remark this answers is not in the document]
Comment (netochaves) [reply test/integration/associations/has-many.test.js]: Sure

**Comment by netochaves** [comment]
No problem @papb, I've added new tests and applied your review.

**Comment by cvharris** [comment]
Is there anyway to get this revived and merged? This would be an awesome feature.

**Comment by WikiRik** [comment]
Closing this since it is still against `master`. Can be reopened if that changes.

**Pull request closed without merging**